### PR TITLE
pinecone init

### PR DESCRIPTION
**Pull request by adivik2000**
This is the early script to upsert the first 10 rows of https://huggingface.co/datasets/Cohere/wikipedia-22-12-simple-embeddings/viewer/Cohere--wikipedia-22-12-simple-embeddings/train dataset and run a sample query based on the content to fetch top 3 results. 

Could you point me to the next steps?

You mentioned that we need an API endpoint to receive the query. What should the response be like?

### CODE REVIEW

File: pine_wiki.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+pinecone_indexes = pinecone.list_indexes()
+
+
+# Step 1: Create an index in Pinecone
+# Define the index name
+index_name = "wikipedia-embeddings"
+
+# Create the index if it doesn't exist
+if index_name not in pinecone_indexes:
+    pinecone.create_index(index_name, dimension=768, metric="euclidean")
```
Comment (NirantK) [inline pine_wiki.py:22]: cosine! Use cosine everywhere!

File: pine_wiki.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        )
+    )
+
+print(vectors)
+
+# Upsert the vectors
+
+# Note: When upserting larger amounts of data, upsert data in batches
+# of 100 vectors or fewer over multiple upsert requests.
+index.upsert(vectors)
```
Comment (NirantK) [inline pine_wiki.py:65]: Can we do this in batches of 100 vectors? i.e. assume a longer dataset and and implement batching

File: pine_wiki.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+query_embeds = co.embed([query], model="multilingual-22-12").embeddings
+
+result = index.query(
+    vector=query_embeds[0],
+    top_k=3,
+    include_values=True,
+    include_metadata=True,
+)
+
+print(result)
```
Comment (NirantK) [inline pine_wiki.py:81]: Thanks for sharing the output, helpful! 

This should be wrapped into a FastAPI endpoint and separated from the upsert/insert. This is our `ask` endpoint. Implement some token based auth such that we don't have to store the token in our front end.

File: pine_wiki.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+# Create the index if it doesn't exist
+if index_name not in pinecone_indexes:
+    pinecone.create_index(index_name, dimension=768, metric="euclidean")
+
+
+# Step 2: Add embeddings to the index
+
+# Load the dataset
+dataset = load_dataset("Cohere/wikipedia-22-12-simple-embeddings", split="train")
```
Comment (NirantK) [inline pine_wiki.py:28]: To begin with, let's use the test during dev. It's smaller and will allow you to run this quickly several times. And during deployment make it easier to deploy.

File: pine_wiki.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+import pinecone
+from datasets import load_dataset
+
+pinecone.init(
+    api_key=os.environ["PINECONE_API_KEY"],
+    environment=os.environ["PINECONE_ENVIRONMENT"],
+)
+
+# List the indexes
+pinecone_indexes = pinecone.list_indexes()
```
Comment (NirantK) [inline pine_wiki.py:13]: Refactor this into an abstract `VectorDB` class. Implement a `Pinecone` class which inherits from `VectorDB` and implements the necessary functions. 

Then write a similar class for say Qdrant Cloud client.

**Review by NirantK — CHANGES_REQUESTED** [review]
In this PR, do the refactor to class, batching and the rest in the next PR please

File: pine_wiki.py (hunk 1 above)
Comment (adivik2000) [reply pine_wiki.py]: This is done.

File: pine_wiki.py (hunk 5 above)
Comment (adivik2000) [reply pine_wiki.py]: Did this. Please take a look?

File: pine_wiki.py (hunk 4 above)
Comment (adivik2000) [reply pine_wiki.py]: As this dataset has only train data, we're going with it.

**Review by NirantK — APPROVED** [review]

**Pull request merged**